### PR TITLE
fix write skew issue

### PR DESCRIPTION
**Pull request by feeblefakie**
It fixes part of https://github.com/scalar-labs/scalardb/issues/65 .

The fix should probably be called workaround since it aborts if it reads empty record. So the write skew case with empty records can not happen in Scalar DB. 
It's a little limitation in the current implementation but I think there seems nothing else we can do since we detect conflicts by using records.

### CODE REVIEW

File: src/integration-test/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitIntegrationTest.java [hunk 1]
```diff
@@ -1189,6 +1195,124 @@ public void commit_NonConflictingDeletesGivenForExisting_ShouldCommitBoth()
     assertThat(another.get(gets.get(account4)).isPresent()).isFalse();
   }
 
+  public void commit_WriteSkewOnExistingRecordsWithSnapshot_ShouldProduceNonSerializableResult()
```
Comment (feeblefakie) [inline src/integration-test/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitIntegrationTest.java:1198]: This is just a new test verifying if write skew actually happens in SNAPSHOT. Just added for confirmation.

File: src/integration-test/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitIntegrationTest.java [hunk 2]
```diff
@@ -1189,6 +1195,124 @@ public void commit_NonConflictingDeletesGivenForExisting_ShouldCommitBoth()
...
+    transaction = manager.start();
+    result1 = transaction.get(get1_1);
+    result2 = transaction.get(get2_1);
+    // the results can not be produced by executing the transactions serially
+    assertThat(result1.get().getValue(BALANCE).get()).isEqualTo(new IntValue(BALANCE, 2));
+    assertThat(result2.get().getValue(BALANCE).get()).isEqualTo(new IntValue(BALANCE, 2));
+  }
+
+  public void
+      commit_WriteSkewOnExistingRecordsWithSerializable_OneShouldCommitTheOtherShouldThrowCommitConflictException()
```
Comment (feeblefakie) [inline src/integration-test/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitIntegrationTest.java:1239]: This is also a new test verifying if write skew actually avoided in SERIALIZABLE with non-empty records. Just added for confirmation.

File: src/integration-test/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitIntegrationTest.java [hunk 3]
```diff
@@ -1189,6 +1195,124 @@ public void commit_NonConflictingDeletesGivenForExisting_ShouldCommitBoth()
...
+    // Assert
+    transaction = manager.start(Isolation.SERIALIZABLE);
+    result1 = transaction.get(get1_1);
+    result2 = transaction.get(get2_1);
+    assertThat(result1.get().getValue(BALANCE).get()).isEqualTo(new IntValue(BALANCE, 1));
+    assertThat(result2.get().getValue(BALANCE).get()).isEqualTo(new IntValue(BALANCE, 2));
+    assertThat(thrown).isInstanceOf(CommitConflictException.class);
+  }
+
+  public void commit_WriteSkewOnNonExistingRecordsWithSerializable_ShouldThrowCommitException()
```
Comment (feeblefakie) [inline src/integration-test/java/com/scalar/db/transaction/consensuscommit/ConsensusCommitIntegrationTest.java:1279]: This is a test to verify the fix is working as expected.
Since there is no way to detect conflicts for empty records correctly, it will abort when it reads empty record with SERIALIZABLE. It is a false positive abort so there might be a case where a history is serializable without such abort.

File: src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java [hunk 4]
```diff
@@ -25,7 +26,7 @@
   private static final Logger LOGGER = LoggerFactory.getLogger(Snapshot.class);
   private final String id;
   private final Isolation isolation;
-  private final Map<Key, TransactionResult> readSet;
+  private final Map<Key, Optional<TransactionResult>> readSet;
```
Comment (feeblefakie) [inline src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java:29]: Now, empty record is also put into the snapshot to identify reading empty records.

File: src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java [hunk 5]
```diff
@@ -117,11 +122,16 @@ void toSerializable() {
                 return;
               }
 
+              if (!e.getValue().isPresent()) {
```
Comment (feeblefakie) [inline src/main/java/com/scalar/db/transaction/consensuscommit/Snapshot.java:125]: Here is the fix.

**Comment by feeblefakie** [comment]
Yonezawa-san is added to a reviewer. 
@Yonezawa-T2 Thank you very much for the report !

**Comment by Yonezawa-T2** [comment]
There is a better fix: creating a record with state `DELETED` in prepare phase prevents the anomary:

- T1: start
- T2: start
- T1: Read ("foo", "t2") and put the value into A. Put 0 if the record does not exist.
- T2: Read ("foo", "t1") and put the value into B. Put 0 if the record does not exist.
- T1: Add 1 to A and put the value to ("foo", "t1")
- T2: Add 1 to B and put the value to ("foo", "t2")
- T1: commit (succeeds)
- T2: commit (fails since T2 cannot create recored ("foo", "t1") with `DELETED` state).

**Comment by feeblefakie** [comment]
@Yonezawa-T2 Thanks ! Yes, that's one of my future plans and is maybe coming.
But we need to make the current delete implementation logical instead of physical which
* causes lots of critical changes in the current code (so it requires lots of verification)
* might not be acceptable by users who manage personal information
so, we will work on after the above are settled.
Logical delete would fix another issue nicely so it should be the from technical perspective.
Thank you, anyways !

**Comment by Yonezawa-T2** [comment]
Prepared records with state `DELETED` are deleted physically in `commitRecords` phase.

https://github.com/scalar-labs/scalardb/blob/f2c7dd4235a7fd481dcfde78c8ef754ccaefae73/src/main/java/com/scalar/db/transaction/consensuscommit/CommitMutationComposer.java#L84-L94

**Comment by yito88** [comment]
@Yonezawa-T2 Thank you for your contribution!
I'm in charge of verification for Scalar DB.

I think it's an impractical change to add the new state of a record as @feeblefakie said.
I would be glad to review and verify your proposal if you make that PR.
However, it will take a few weeks to make a big change pass our verification.

I would like to avoid this issue with the small change for now.

**Comment by feeblefakie** [comment]
@Yonezawa-T2 Yes, writing DELETE would work better when thinking about only this issue. But I think introducing logical delete would be better when thinking about #65 and #66 . So, even if writing DELETE here, the code will be updated, so I want to keep this change as minimum as possible for safety.

**Comment by Yonezawa-T2** [comment]
@yito88 `DELETED` state is not a new state. It is used to delete a existing record.

https://github.com/scalar-labs/scalardb/blob/f2c7dd4235a7fd481dcfde78c8ef754ccaefae73/src/main/java/com/scalar/db/transaction/consensuscommit/PrepareMutationComposer.java#L79-L105

**Comment by yito88** [comment]
@Yonezawa-T2 Sorry it's not adding the new state as you said. It's adding a new state transition.
